Fundamentals metadata

RJTK opened the pull request:
🎉 Hello there 🎉! Thanks for taking the time to contribute to `Tiingo-Python`!

## Summary of Changes

- What is the motivation / context for this change
   -  There is a bunch of metadata available at the url tiingo/fundamentals/metadata, the api should have a method to query it
- Is this a bugfix or a feature?
- 
- What does this change / fix? (Link to Github Issue page if applicable)

## Checklist

- [ ] Code follows the style guidelines of this project
- [ ] Added tests for new functionality
- [ ] Update `HISTORY.rst` with an entry summarizing your change
- [ ] Tag a project maintainer

hydrosquall commented:
Hey @RJTK ,

Thanks so much for this contribution. I think you're right that 2.6 and 2.7 are causing issues and it's not worth the maintenance burden given that 2 was sunset at the end of last year.

I also use `black` on new projects, but started this project before I knew that existed. Once we are fully into python 3, I think it would be good to both do a 1 time pass over the whole codebase with black, and then add it to CI.

I think this will also let us simplify the parts of the code that check for whether users are running python 2 or 3.

Let me open up an issue to sunset Python 2, with some of the checklist items that are needed for us to switch - I'd be happy to review any PRs you have time for to help with that effort, otherwise I'll try to look into this more on the weekend.

hydrosquall commented:
Opened https://github.com/hydrosquall/tiingo-python/issues/601 with a shortlist of things needed to phase out python 2.

RJTK commented:
Cool thanks -- I will look at #601 and see if I can address any of the points.  I am not well versed in these integration tools, but would probably be well advised to learn.

hydrosquall commented:
@RJTK I did some cleanup this weekend, can you try rebasing your branch against the latest `master`? The CI should be better than before now, and we've dropped support below python 3.6.

> Perhaps this method should also just replace or be combined with get_ticker_metadata? However, they are two separate endpoints.

My concern is that the two endpoints may not have identical signatures, and we could end up with parameters in the function that are only meant to be used by one or the other. For the moment, I think it's OK for us to keep the pattern of roughly 1 endpoint per function for these metadata calls to keep things interpretable, even though the get_ticker_price is not really following this rule (since it uses a different endpoint for intraday data).

> I cleaned up a bunch of repeated code by adding _format_response

Brilliant, thank you!

> I cache the list of available tickers in _ticker_list so that all the list_tickers functions don't collect the zip file multiple times.

Ah, thanks for looking into this optimization. What are your thoughts on what a user should do if they want to invalidate the cached results?

RJTK commented:
> roughly 1 endpoint per function

This makes sense to me.

> Ah, thanks for looking into this optimization. What are your thoughts on what a user should do if they want to invalidate the cached results?

This didn't occur to me -- I added an option to force it to refresh the cache and updated some comments to make it clear that the results were cached.  We could also just get rid of this caching -- my assumption was that it wouldn't be asking the website for the csv file each time the method was called, and that it was basically just a "getter" on a class attribute -- but I realize now that the opposite assumption may also be natural.

I think some tests are failing on the CI because there's no API key?  How do I update the fixtures without committing my API key to git?  Just set the key to 00000000... ?

hydrosquall commented:
Hi @RJTK , we use a project called `vcr.py` so you can record the JSON response locally, and push that response into the `tests/fixtures` directory.
https://github.com/hydrosquall/tiingo-python/pull/32

Once you have your fixture, you can use the API key cleaning tool introduced in to replace your API key from the response with the `0000...` dummy API key,.

https://github.com/hydrosquall/tiingo-python/pull/107

The discoverability of our docs/test-writing structure can be improved, but for now everything you need should be in the tips section: https://github.com/hydrosquall/tiingo-python/blob/master/CONTRIBUTING.rst#tips .

> We could also just get rid of this caching -- my assumption was that it wouldn't be asking the website for the csv file each time the method was called, and that it was basically just a "getter" on a class attribute -- but I realize now that the opposite assumption may also be natural.

We've erred on the side of not caching so as to avoid returning stale results. If we do have caching as an option, I'd encourage us towards having it be disabled by default so that anyone who currently expects to get fresh results on every call won't have to edit their programs to use the new parameter (this ensures backwards compatibility).

Let me know if I can help with anything else. Thanks for following up on this!

RJTK commented:
I removed caching for the ticker symbols and commit the test fixtures.

hydrosquall commented on tiingo/api.py:
```diff
@@ -85,59 +86,55 @@ def __init__(self, *args, **kwargs):
-           https://apimedia.tiingo.com/docs/tiingo/daily/supported_tickers.zip
-           """
+    @staticmethod
+    def _get_ticker_list():
```
We may not gain a lot in terms of readability by having this separate function when there isn't caching, and in fact we'll pay a small performance price by initializing the list twice (both here and inside `list_tickers`, whereas before that was only created once. 

I'd suggest either replacing this list comprehension with a generator comprehension ( https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Generators_and_Comprehensions.html#Creating-your-own-generator:-generator-comprehensions ), or letting this part just be part of the original `list_tickers` to avoid the duplicate work.

hydrosquall reviewed (changes requested):
Hi @RJTK , this is looking good, thanks for looking into VCR.py , the CI is nice and 🟢 now. Now that they're passing, I was able to take a closer look at the actual implementation. Let me know if you have any questions.

hydrosquall commented:
Hi @RJTK , just wanted to check on the status of your PR / if you had any questions.

RJTK commented:
Hi yes, sorry -- I actually just completely forgot about this since I had what I needed working and moved onto another aspect of the project.  Thanks a lot for the feedback -- I will update the PR by tomorrow.

hydrosquall commented:
No rush, and glad to hear things worked out in your project :)! Thanks for
taking the time to contribute upstream rather than just updating your local
copy, it’s much appreciated (especially since it probably ended up adding
more steps than originally anticipated).

On Sat, Apr 10, 2021 at 5:44 PM RJTK ***@***.***> wrote:

> Hi yes, sorry -- I actually just completely forgot about this since I had
> what I needed working and moved onto another aspect of the project. Thanks
> a lot for the feedback -- I will update the PR by tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hydrosquall/tiingo-python/pull/596#issuecomment-817206260>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACE2MM4EAQDZVCTYI42OGCTTIDBCNANCNFSM4YGYO3KQ>
> .
>

RJTK commented on tiingo/api.py:
```diff
@@ -85,59 +86,55 @@ def __init__(self, *args, **kwargs):
-                if row.get('assetType') in assetTypesSet]
+    @staticmethod
+    def _format_response(response, fmt, object_name=None):
+        """Converts the api response into json, csv, or objects.
```
Not sure I understand this comment.  The `object_name` cannot be a list.  Or, do you mean to clarify that the returned result can be either a single namedtuple or list thereof?

hydrosquall replied:
Ah the positioning of my comment was unfortunate, I was indeed referring to the type of `objects` in the docstring rather than the object_name parameter. Your interpretation of `mean to clarify that the returned result can be either a single namedtuple or list thereof` is right :).

RJTK commented:
What does it mean to run a workflow?  From the docs it says that "Forks of public repositories can submit pull requests that propose changes to a repository's GitHub Actions workflows" -- have I modified a file that affects something here?

I changed the test tickers to PG and CAT -- these are part of the DOW30 and seem to be more freely available, so users that don't have or don't have yet premium keys will hopefully have fewer problems.

hydrosquall commented:
> What does it mean to run a workflow? From the docs it says that "Forks of public repositories can submit pull requests that propose changes to a repository's GitHub Actions workflows" -- have I modified a file that affects something here?

I think there was a recent change to how github actions work due to issues with cryptocurrency mining abuse. I've clicked 'approve', so you shouldn't have any issues with this.

https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/

For future reference, running a workflow just means running the commands in https://github.com/hydrosquall/tiingo-python/blob/master/.github/workflows/python-package.yml, which happens based on the events specified in these lines:

https://github.com/hydrosquall/tiingo-python/blob/ff845a249a208718a88d4a3c927980b1b7ebf06e/.github/workflows/python-package.yml#L6-L10

> I changed the test tickers to PG and CAT 

Ah, thank you! FWIW generally users shouldn't need to have a tiingo api account (or even internet connection) to be able to run the unit tests, this is the purpose of recording the API responses as fixtures. With that said, I don't have particular attachment to GOOG or APPL, so switching to these tickers is fine with me. 

From the looks of it, there might be some additional tests to update 

https://github.com/hydrosquall/tiingo-python/pull/596/checks?check_run_id=2487772634

The intraday pricing cassette that was re-recorded also needs to have its ticker updated from GOOG since it's used by the `pandas` test suite too.

https://github.com/hydrosquall/tiingo-python/blob/ff845a249a208718a88d4a3c927980b1b7ebf06e/tests/test_tiingo_pandas.py#L110-L116

hydrosquall commented:
Hi @RJTK - just doing some project maintenance and wanted to see if you might have the chance to look into the current test issues . My guess is that since you changed the test tickers, the old cassette fixtures will need to be regenerated.
